lib/src/rules/avoid_dynamic_calls.dart

```diff
@@ -0,0 +1,226 @@
+      ..addMethodInvocation(this, visitor)
+      ..addPostfixExpression(this, visitor)
+      ..addPrefixExpression(this, visitor)
+      ..addPrefixedIdentifier(this, visitor)
+      ..addPropertyAccess(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  bool _lintIfDynamic(Expression node) {
+    if (node?.staticType?.isDynamic == true) {
+      rule.reportLint(node);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  void _lintIfDynamicOrFunction(Expression node) {
+    final staticType = node?.staticType;
+    if (staticType == null) {
+      return;
+    }
+    if (staticType.isDynamic || staticType.isDartCoreFunction) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  void visitAssignmentExpression(AssignmentExpression node) {
+    if (node.readType?.isDynamic != true) {
+      return;
+    }
+    if (node.operator.type == TokenType.QUESTION_QUESTION_EQ) {
+      // x ??= foo is not a dynamic call.
+      return;
+    }
+    rule.reportLint(node);
+  }
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    switch (node.operator.type) {
+      case TokenType.QUESTION_QUESTION:
+        return;
+      case TokenType.EQ_EQ:
+      case TokenType.BANG_EQ:
+        if (node.rightOperand is NullLiteral) {
+          return;
+        }
+        break;
+    }
+    if (node.operator.type == TokenType.AS ||
+        node.operator.type == TokenType.IS) {
+      // Can't be included in switch statement due to Dart semantics.
+      return;
+    }
+    _lintIfDynamic(node.leftOperand);
+    // We don't check node.rightOperand, because that is an implicit cast, not a
+    // dynamic call (the call itself is based on leftOperand). While it would be
+    // useful to do so, it is better solved by other more specific lints to
+    // disallow implicit casts from dynamic.
+  }
+
+  @override
+  void visitFunctionExpressionInvocation(FunctionExpressionInvocation node) {
+    if (node.function != null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitIndexExpression(IndexExpression node) {
+    _lintIfDynamic(node.realTarget);
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    final receiverWasDynamic = _lintIfDynamic(node.realTarget);
+    if (!receiverWasDynamic && node.target == null) {
+      _lintIfDynamicOrFunction(node.function);
+    }
+  }
+
+  @override
+  void visitPrefixExpression(PrefixExpression node) {
+    if (_lintIfDynamic(node.operand)) {
+      return;
+    }
+    switch (node.operator.type) {
+      case TokenType.PLUS_PLUS:
+      case TokenType.MINUS_MINUS:
+        // The ++ and -- operator expressions don't resolve a static type.
+        if (node.staticType?.isDynamic == true) {
+          rule.reportLint(node.operand);
+        }
+    }
+  }
+
+  @override
+  void visitPrefixedIdentifier(PrefixedIdentifier node) {
```

Comment:
The primary use case for `PrefixedIdentifier` is to represent an identifier that's preceded by an import prefix. In those cases we don't want to trigger the lint. It's quite possible that `_lintIfDynamic` will correctly not report a lint in those cases, but we should add a test to ensure that there isn't a false positive here. Perhaps something like
```
import 'dart:math' as math;

var p = math.pi;
```

Reply:
Woof. This was tricky, @srawlins helped a lot. The (mock?) sdk is not setup to import `dart:math` in these test cases, so I went with `dart:core` and `print`. Hope that's fine

Reply:
I didn't realize that `dart:math` wasn't supported, sorry. But yes, any valid import works. I just wanted to ensure that prefixes didn't look like they had the type `dynamic`.